### PR TITLE
Fix for #Pydev 1106: Exception during pretty-printing

### PR DESCRIPTION
**Pull request by oluiscabral**

### CODE REVIEW

File: plugins/org.python.pydev.parser/src/org/python/pydev/parser/prettyprinterv2/MakeAstValidForPrettyPrintingVisitor.java [hunk 1]
```diff
@@ -1085,6 +1088,13 @@ public Object visitExpr(Expr node) throws Exception {
         return null;
     }
 
+    @Override
+    public Object visitExtSlice(ExtSlice node) throws Exception {
+        fixNode(node);
```
Comment (fabioz) [inline plugins/org.python.pydev.parser/src/org/python/pydev/parser/prettyprinterv2/MakeAstValidForPrettyPrintingVisitor.java:1093]: The proper fix isn't to do the traverse but rather override it and then say what's required at each place.

So, you'd do code to iterate through `ExtSlice.dims` and then add `doc.addRequire(",", lastNode);` after each `dim` which you iterated (close to what the `org.python.pydev.parser.prettyprinterv2.PrettyPrinterVisitorV2.visitImport(Import)` does when handling the import names).

File: plugins/org.python.pydev.parser/src/org/python/pydev/parser/jython/ast/Slice.java [hunk 2]
```diff
@@ -63,6 +64,12 @@ public Slice createCopy(boolean copyComments) {
                 if(o instanceof commentType){
                     commentType commentType = (commentType) o;
                     temp.getSpecialsAfter().add(commentType.createCopy(copyComments));
+                } else if (o instanceof SpecialStr) {
```
Comment (fabioz) [inline plugins/org.python.pydev.parser/src/org/python/pydev/parser/jython/ast/Slice.java:67]: This should not be needed... also, if this was ever needed you'd have to change `asdl_java.py` to generate the code for you (the AST code is all generated).

File: plugins/org.python.pydev.parser/src/org/python/pydev/parser/jython/ast/Index.java [hunk 3]
```diff
@@ -51,6 +52,12 @@ public Index createCopy(boolean copyComments) {
                 if(o instanceof commentType){
                     commentType commentType = (commentType) o;
                     temp.getSpecialsAfter().add(commentType.createCopy(copyComments));
+                } else if (o instanceof SpecialStr) {
```
Comment (fabioz) [inline plugins/org.python.pydev.parser/src/org/python/pydev/parser/jython/ast/Index.java:55]: Revert here too.

File: plugins/org.python.pydev.parser/src/org/python/pydev/parser/prettyprinterv2/MakeAstValidForPrettyPrintingVisitor.java [hunk 4]
```diff
@@ -133,6 +134,8 @@ protected void fixAfterNode(SimpleNode node) {
                 if (o instanceof commentType) {
                     fixNode((SimpleNode) o);
                     nextLine();
+                } else if (o instanceof Name) {
```
Comment (fabioz) [inline plugins/org.python.pydev.parser/src/org/python/pydev/parser/prettyprinterv2/MakeAstValidForPrettyPrintingVisitor.java:137]: Revert here.

**Review by fabioz — CHANGES_REQUESTED** [review]

File: plugins/org.python.pydev.parser/src/org/python/pydev/parser/prettyprinterv2/MakeAstValidForPrettyPrintingVisitor.java (hunk 1 above)
Comment (oluiscabral) [reply plugins/org.python.pydev.parser/src/org/python/pydev/parser/prettyprinterv2/MakeAstValidForPrettyPrintingVisitor.java:1093]: Let me see if I could understand it.

So we should just return null with ExtSlice nodes (without traversing it nor using fixNode) and then visit it in PrettyPrinterVisitorV2, adding commas in between the dims?

(the exception throwed in this issue was about not having a ExtSlice visitor in MakeAstValidForPrettyPrintingVisitor)

Comment (fabioz) [reply plugins/org.python.pydev.parser/src/org/python/pydev/parser/prettyprinterv2/MakeAstValidForPrettyPrintingVisitor.java:1093]: Yes (my comments there were related to the `PrettyPrinterVisitorV2.visitExtSlice`). 

On the `MakeAstValidForPrettyPrintingVisitor`, you also need to override `visitExtSlice`, but here I think it's something like:

```
    @Override
    public Object visitExtSlice(ExtSlice node) throws Exception {
        fixNode(node);
        if(node.dims != null) {
            for (int i = 0; i < node.dims.length; i++) {
                node.dims[i].accept(this);
                nextCol();
            }
        }
        fixAfterNode(node);
        return null;
    }
```

**Pull request merged**